File: 2019/day-02/lua/part2.lua
local OPADD = 1
local OPMULTIPLY = 2
local OPHALT = 99

function table.clone(org)
    return {table.unpack(org)}
end

function interpret_intcode(noun, verb, num)

    num[2] = noun
    num[3] = verb

    local i = 1

    while true  do
        one = num[i + 1]+1
        two = num[i + 2]+1
        three = num[i + 3]+1
        if num[i] == OPADD then
            num[three] = num[one] + num[two]
            i = i + 4
        elseif num[i] == OPMULTIPLY then
            num[three] = num[one] * num[two]
            i = i + 4
        elseif num[i] == OPHALT then
            break
        end
    end
    return num[1]
end

function find_value()
    local num = {}
    for str in io.lines(arg[1])():gmatch("%d+") do
        num[#num+1] = tonumber(str)
    end
    for noun = 0, 99, 1 do
        for verb = 0, 99, 1 do
            if interpret_intcode(noun, verb, table.clone(num)) == 19690720 then
                return 100 * noun + verb
            end
        end
    end
end

print("Solution: " .. find_value())

File: 2019/day-02/python/part2.py
#!/usr/bin/env python

import sys
from copy import deepcopy

OPADD = 1
OPMULTIPLY = 2
OPHALT = 99

def interpret_intcode(noun, verb, num):
    i = 0

    num[1] = noun
    num[2] = verb

    while True:
        one = num[i + 1]
        two = num[i + 2]
        three = num[i + 3]
        if num[i] == OPADD:
            num[three] = num[one] + num[two]
            i += 4
        elif num[i] == OPMULTIPLY:
            num[three] = num[one] * num[two]
            i += 4
        elif num[i] == OPHALT:
            break
        else:
            return("?")
    return num[0]

def find_value():
    num = [int(i) for i in open(sys.argv[1], "r").read().split(',')]
    for noun in range(0, 100, 1):
        for verb in range(0, 100, 1):
            if interpret_intcode(noun, verb, deepcopy(num)) == 19690720:
                return 100 * noun + verb

print("Solution: " + str(find_value()))